service/src/main/java/bio/terra/tanagra/db/AllArtifactsDao.java

```diff
@@ -0,0 +1,190 @@
+import java.util.Map;
+import org.apache.commons.lang3.tuple.Pair;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.jdbc.core.namedparam.MapSqlParameterSource;
+import org.springframework.stereotype.Component;
+
+// A Dao for all artifacts: this is needed to avoid circular dependency errors
+@Component
+public class AllArtifactsDao {
+  private final CohortDao cohortDao;
+  private final ReviewDao reviewDao;
+  private final AnnotationDao annotationDao;
+
+  @Autowired
+  public AllArtifactsDao(CohortDao cohortDao, ReviewDao reviewDao, AnnotationDao annotationDao) {
+    this.cohortDao = cohortDao;
+    this.reviewDao = reviewDao;
+    this.annotationDao = annotationDao;
+  }
+
+  /**
+   * Clones a cohort and all related artifacts by copying DB rows. Does not include validations.
+   * Returns the id of the new cohort
+   */
+  @WriteTransaction
+  public String cloneCohort(
+      String cohortId,
+      String userEmail,
+      String destinationStudyId,
+      String displayName,
+      String description) {
+    // Cohort
+    Cohort original = cohortDao.getCohort(cohortId);
+    String clonedCohortId = ServiceUtils.newArtifactId();
+
+    cohortDao.insertCohortRow(
+        destinationStudyId,
+        clonedCohortId,
+        original.getUnderlay(),
+        userEmail,
+        displayName,
+        description);
+
+    // Reviews
+    List<Review> originalReviews = reviewDao.getAllReviews(cohortId, 0, Integer.MAX_VALUE);
+    Map<String, String> originalRevisionIdreviewIdMap = new HashMap<>(); // revisionId -> reviewId
+    Map<String, String> reviewIdMap = new HashMap<>(); // originalId -> clonedId
+    List<MapSqlParameterSource> reviewParamSets = new ArrayList<>();
+    List<MapSqlParameterSource> primaryEntityParamSets = new ArrayList<>();
+
+    originalReviews.forEach(
+        review -> {
+          String originalId = review.getId();
+          String clonedId = ServiceUtils.newArtifactId();
+          originalRevisionIdreviewIdMap.put(review.getRevision().getId(), originalId);
+          reviewIdMap.put(originalId, clonedId);
+
+          reviewParamSets.add(
+              reviewDao.buildReviewParam(
+                  clonedCohortId,
+                  clonedId,
+                  review.getSize(),
+                  review.getDisplayName(),
+                  review.getDescription(),
+                  userEmail));
+
+          Map<Literal, Integer> primaryEntityIdsToStableIndex =
+              reviewDao.getPrimaryEntityIdsToStableIndex(originalId);
+          primaryEntityIdsToStableIndex.forEach(
+              (entityLiteral, index) ->
+                  primaryEntityParamSets.add(
+                      reviewDao.buildPrimaryEntityParam(
+                          clonedId, entityLiteral.getInt64Val(), index)));
+        });
+    reviewDao.insertReviewRows(reviewParamSets); // fk: cohort(id)
+    reviewDao.insertPrimaryEntityRows(primaryEntityParamSets); // fk: review(id)
```

Comment:
check: use a direct sql statement

Reply:
Avoiding this for now to use the existing DB functions. helps avoid extra code maintenance if there are changes in the future